src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/MissingOverrideCheck.java

```diff
@@ -250,36 +252,50 @@ public void visitToken(final DetailAST ast) {
 
             if (check
                 && containsTag
-                && !AnnotationUtil.containsAnnotation(ast, OVERRIDE)
-                && !AnnotationUtil.containsAnnotation(ast, FQ_OVERRIDE)) {
+                && !hasOverrideAnnotation(ast)) {
                 log(ast, MSG_KEY_ANNOTATION_MISSING_OVERRIDE);
             }
         }
     }
 
     /**
-     * Checks to see if the text block contains a inheritDoc tag.
+     * Checks to see if the ast has Override annotation.
      *
-     * @param javadoc the javadoc of the AST
-     * @return true if contains the tag
+     * @param ast AST node
+     * @return true if has Override annotation
      */
-    private static boolean containsJavadocTag(final TextBlock javadoc) {
-        boolean javadocTag = false;
-
-        if (javadoc != null) {
-            final String[] lines = javadoc.getText();
-
-            for (final String line : lines) {
-                final Matcher matchInheritDoc =
-                    MATCH_INHERIT_DOC.matcher(line);
+    private static boolean hasOverrideAnnotation(DetailAST ast) {
+        return AnnotationUtil.containsAnnotation(ast, OVERRIDE)
+                || AnnotationUtil.containsAnnotation(ast, FQ_OVERRIDE);
+    }
 
-                if (matchInheritDoc.find()) {
-                    javadocTag = true;
-                    break;
-                }
-            }
+    /**
+     * Checks to see if the ast contains a inheritDoc tag.
+     *
+     * @param ast AST node
+     * @return true if contains the tag
+     */
+    private static boolean containsJavadocTag(DetailAST ast) {
+        final DetailAST modifiers = ast.getFirstChild();
+        final DetailAST startNode;
+        if (modifiers.hasChildren()) {
+            startNode = TokenUtil.findFirstTokenByPredicate(ast.getFirstChild(),
+                    node -> node.getType() == TokenTypes.ANNOTATION)
+                .orElse(modifiers);
+        }
+        else {
+            startNode = ast.findFirstToken(TokenTypes.TYPE);
         }
-        return javadocTag;
+        final Optional<String> javadoc =
+            Stream.iterate(startNode.getLastChild(), Objects::nonNull,
+                    DetailAST::getPreviousSibling)
+            .filter(node -> node.getType() == TokenTypes.BLOCK_COMMENT_BEGIN)
+            .map(DetailAST::getFirstChild)
+            .map(DetailAST::getText)
+            .filter(JavadocUtil::isJavadocComment)
+            .findFirst();
+        return javadoc.isPresent()
+                && MATCH_INHERIT_DOC.matcher(javadoc.get()).find();
```

Comment:
Do you think it is reasonable to avoid regexp here  and do this by extra tree traversing ? We should avoid regexp parsing as much a possible as it is fragile in numerous cases.
it can be separate issue, but lets create it.

Reply:
As I see in AST, block comment text has no children, so we use regex to check its text.
Or you mean some other way?

Reply:
Usage of javadoc ast is possible if we change parent class of Check to be javadoc Check. But it will be performance issue, as all javadoc Check are very slow ( in comparison to java Checks).

For now and better to keep regexp.